### DISCUSSION TITLE
PocketPages 0.9.5 released and 0.10.0-next on the horizon

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5922] [Show and tell]
For anyone following, I made https://pocketpages.dev as an answer for slim-client server-side pages in PocketBase. I see a not-so-distant future where I can dump the heavy frontend build tooling and slim down to htmx+pocketpages.

Here's what PocketPages looks like:

```
// pb_hooks/pages/index.ejs
<%= `Hello, world!` %>
```

## Available Starters

- **minimal** - One file, zero deps beyond PocketPages
- **htmx** - HTMX integration for dynamic updates
- **mvp** - MVP.css for quick prototypes
- **daisyui** - Tailwind + DaisyUI components

Probably htmx is the most interesting starter, check it out:

```bash
cp -r node_modules/pocketpages/starters/htmx .
cd htmx
npm i
npm run dev
```

A big upgrade to v0.10 is on the horizon with breaking changes. If you want a preview of that, you can look at https://pocketpages.dev/docs-next.

**Comment by dschissler** [comment]
I'd like to give an update on my latest work for making single file shadow dom components.  I was told not to spam.  So this will be my last post until quite a ways from now when I'm ready to release this with pretty good documentation and examples.  Basically, I've solved shadow dom single file web components.  The total unminified single file library with Alpine and Htmx support is currently only 14kb (plus Alpine and Htmx libs themselves.).


```html
<div class="grid">
  <button @click="count--">(-)</button>
  <button @click="dec">-</button>
  <h2 id="count" x-text="count"></h2>
  <button @click="inc">+</button>
</div>

<script setup="alpine">
  export default class extends HTMLElement {
    static get props() {
      return {
        count: {
          default: 0,
          cast: parseInt,
        }
      }
    }

    static get data() {
      return {
        dec() {
          this.count--
        },
        inc() {
          this.count++
        },
      }
    }

    connectedCallback() {
      this.$effect(() => {
        console.log('count: ', this.count, typeof this.count)
      })
    }
  }
</script>

<style>
  /* You can use ids because it's shadow dom */
  #count {
    justify-self: center;
    align-self: center;
  }
</style>
```

For declarative functionality to define components set it up like this:

```html
  <script type="importmap">
    {
      "imports": {
        "unipow": "/unipow/unipow.js",
        "alpinejs": "/unipow/alpine.js",
        "htmx.org": "/unipow/htmx.js"
      }
    }
  </script>
  <script
    upow
    type="module"
    src="/unipow/unipow.js">
  </script>
```

Then define components using a template list.  You can declare the prefix and postfix (directory and extension) at the global level and at the declaration (import) level.

```html
<template upow list>
  <li>counter-tool</li>
</template>
```

Or using a template single definition:

```html
<template upow one="counter-tool"></template>
```

Or define it inline:

```html
<template upow name="this-is-a-custom-element">
  This can now be used as a shadow dom custom element anywhere.  You can add markup, script, or style to it at your leisure.
</template>

<this-is-a-custom-element></this-is-a-custom-element>
```

Or just do it all via script:

```javascript
  <script type="module">
    const {Unipow, UnipowAlpine} = await import('/unipow/unipow.js')
    const unipow = new Unipow
    unipow.plugins.load('alpine', UnipowAlpine)

    unipow.list([
      'web-cam',
    ])
    unipow.start()
  </script>
```

You can also create a self-instantiating inline (shadow dom) web component for things like layout or one-offs.  This allows registered plugin frameworks (htmx and alpine for now) are able to be provided the functionality at a local level.  So htmx and alpine is not global, but rather is setup and updated within the shadow dom.  It works!

```html
<template upow now>
  <div hx-boost="true" hx-target="global #main" hx-select="#main" hx-swap="outerHTML">
    <a href="/">Home</a>
    <a href="/webcams">Webcams</a>
    <a href="/testing">testing</a>
  </div>

  <script setup="alpine,htmx">
  </script>
</template>
```

It even has full mutation observer support:

```html

{{define "title"}}
Mutation Observer Test for custom element defined by template.
{{end}}

{{define "body"}}

<script type="module">
  setTimeout(() => {
    const tpl = document.createElement('template')
    tpl.setAttribute('upow', '')
    tpl.setAttribute('name', 'mut-test')
    tpl.innerHTML = 'Mutation Test.'
    document.querySelector('main')?.appendChild(tpl)
  }, 400)
</script>

<mut-test></mut-test>

{{end}}
```

I've solved styles by allowing you to put `upow` on any link or style element in the light dom.  This will then be recreated as a constructable style sheet that is attached to each custom element upon instantiation.

```html
  <link upow rel="stylesheet" href="/lib/style/chota-0.8.1.min.css">
```


On release I'll also have petite-vue support, and possibly some other minor ones.  I'm quite curious to see if petite-vue and alpine components will work together for the reactive property binding since they both use the vue reactive library.  I mean that anything defined as `static props ...` in the web component is available in a `v-for` / `x-for` loop just as any other element.  It's just a matter of if these two libraries have completely compatible reactivity.

So that's it for some months until I release this buildless, single-file, normally-styllable, single-component library that can demand plugins on a component-by-component basis.

**Reply by benallfree** [reply]
Can you please explain how this relates to PocketPages?

**Reply by dschissler** [reply]
It seems that you're looking for a way to focus on Pocketbase functionality in a way that allows you to reduce or eliminate the need of server-side building of client resources.  Just saying "htmx" alone implies that you want to be able to reduce a lot of boilerplate for mundane things.

What happens with a buildless system is that there is a giant chasm between normal sprinkled Alpine and more advanced functionality.  Lit Element is awkward and feels out of place.  The same with Preact, etc.  It's much better if a template tag can be produced as html via the server-side rendering, or imported in javascript via html declarations.  That is the motivation behind declarative shadow dom (DSD), but doesn't really do much at the moment, and it requires another copy per component (because it doesn't register anything).

[edit] Apparently you can define elements into the registry in DSD.

**Reply by dschissler** [reply]
I suppose that I posted this more so for how you might be able to read the tea leaves as you continue developing, not so much that you'd use my framework.  I think that the future for this type of site is Svelte/Vue type single file components that can be delivered without a build step, and in a way that allows them to be included inline server-side as well as imported in the client.

**Reply by benallfree** [reply]
I see, thank you.

Yes, PocketPages would be a good fit for rendering Web Components or some other type of homebrew component architecture (which is what it seems like you're proposing?). Is that right?

**Reply by dschissler** [reply]
Yes, that's right.  I'm convinced that I'm on the right path, but of course the taste is all in the implementation.  Alpine integration seems to be the best fit at the moment due to the large community support, but I feel that a slightly enhanced (and not abandoned) petite-vue is the better fit.  It's just far beyond my R&D budget to get into directly working on libraries like these.

I think that for the moment piggy-backing on Alpine is ideal because then I'm not asking anyone to learn something very weird.  There is really only `static props` to define props (and there is only three settings for each, two of which are optional), and `static data` function to define a reactive store that is usable in the component and is also available in the alpine-enabled template.

I mean to say that the custom-isms is a very tiny thing to learn, and then the rest of it is all transferable to any other project (alpine, htmx and custom elements themselves).  So I mean to say is that ideally there is just a tiny opinionated wrapper around custom elements, but that custom elements are still exposed and not written away.

I watched a podcast with Evan You of Vue.  I took some of what he said about petite-vue to heart.  He talked about the limited scope, and not needing to creep it to some use cases that are mostly irrelevant due to petite-vue never being a good fit for things that need to be very complicated or efficient.  So I mean to say that I deleted a bunch of features from my library because I determined that nobody should ever care about them.  If they really care then they should go big with the monstrosity of a node based build system.
